### ISSUE TITLE
Schema building dsl with configuration

### ISSUE THREAD

**Issue by johnhungerford** [johnhungerford/generic-schema#22]
```
trait MyCustomConfiguration extends SchemaConfiguration {
    override def configure : Configuration = ???
}

...

import MyCustomConfiguration.dsl.*

val schema = 
    schema[ SomeType ]
        .product
        .addField(field[NestType].fieldName("nested_field").build)
        ...
        .updateComponent("nested_field"/"int_field")(_.withName("int"))
        .updateComponent("nested_field"/"str_field")(_.updateSchema(_.genericName("string")))))
        .build
```

Standardize methods: 
* `replace` replaces with new (for build only)
* `update` modify via Schema[T] => Schema[T]
* `rebuild` modify via SchemaBuilder[T] => Schema[T]

**Comment by johnhungerford** [comment]
* Schema usage should be possible without any imports (e.g., Schema type should not be nested, and Schema methods should not require importing given instances
* Schema building, modifying, and rebuilding, should require `.dsl.{*, given}` imports (e.g., someSchema.updateField method should not be visible without an extension)
* Corollary: configuration instance should not be necessary for schema usage

**Comment by johnhungerford** [comment]
Another possible pattern
```
val schema = MyCustomConfiguration.usingDsl { dsl =>
    import dsl.{*, given}

    schema[ SomeType ]
        .product
        .addField(field[NestType].fieldName("nested_field").build)
        ...
        .updateComponent("nested_field"/"int_field")(_.withName("int"))
        .updateComponent("nested_field"/"str_field")(_.updateSchema(_.genericName("string")))))
        .build
}
```

**Comment by johnhungerford** [comment]
Builder (only available with extensions in scope):
```
Schema.buildProduct[T]
Schema.buildCoproduct[T]
Schema.buildDerived[T]
Schema.buildPrimitive[T]

Schema.derived[T]  // no need to call .build
Schema.primitive[T](name, description, validators) // no need to call .build
```

**Comment by johnhungerford** [comment]
Closed with #29
